File: array/problem51/problem51.py
from typing import List


class Solution:
    def solveNQueens(self, n: int) -> List[List[str]] :
        full = (1 << n) - 1
        queens: List[int] = []
        result: List[List[str]] = []

        def backtrack(cols: int, d1: int, d2: int) -> None:
            if cols == full:
                board = ["." * q + "Q" + "." * (n - q - 1) for q in queens]
                result.append(board)
                return
            avail = full & ~(cols | d1 | d2)
            while avail:
                b = avail & (-avail)          # isolate lowest set bit
                avail ^= b
                col = b.bit_length() - 1
                queens.append(col)
                backtrack(cols | b, (d1 | b) >> 1, (d2 | b) << 1)
                queens.pop()

        backtrack(0, 0, 0)
        return result

File: array/problem51/problem51.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
    int n_, full_;
    vector<int> queens_;
    vector<vector<string>> result_;

    void bt(int cols, int d1, int d2) {
        if (cols == full_) {
            vector<string> board(n_, string(n_, '.'));
            for (int r = 0; r < n_; ++r) board[r][queens_[r]] = 'Q';
            result_.push_back(move(board));
            return;
        }
        int avail = full_ & ~(cols | d1 | d2);
        while (avail) {
            int b    = avail & (-avail);        // lowest set bit
            avail   &= avail - 1;
            int col  = __builtin_ctz(b);
            queens_.push_back(col);
            bt(cols | b, (d1 | b) >> 1, (d2 | b) << 1);
            queens_.pop_back();
        }
    }

public:
    vector<vector<string>> solveNQueens(int n) {
        n_ = n;
        full_ = (1 << n) - 1;
        queens_.clear();
        result_.clear();
        bt(0, 0, 0);
        return result_;
    }
};
